akka-persistence/src/main/scala/akka/persistence/journal/WriteJournalBase.scala

```diff
@@ -4,21 +4,36 @@
 
 package akka.persistence.journal
 
-import akka.persistence.{ PersistentRepr, PersistentEnvelope }
+import akka.persistence.{ Persistence, PersistentRepr, PersistentEnvelope }
 import akka.actor.Actor
 import scala.collection.immutable
 
 private[akka] trait WriteJournalBase {
   this: Actor ⇒
 
+  private lazy val eventAdapters = Persistence(context.system).adaptersFor(context.self)
+
   protected def preparePersistentBatch(rb: immutable.Seq[PersistentEnvelope]): immutable.Seq[PersistentRepr] =
-    rb.filter(persistentPrepareWrite).asInstanceOf[immutable.Seq[PersistentRepr]] // filter instead of flatMap to avoid Some allocations
+    rb.collect {
+      case p: PersistentRepr ⇒
+        p.prepareWrite()
+        adaptToJournal(p)
+    } // collect instead of flatMap to avoid Some allocations
+
+  /** INTERNAL API */
+  private[akka] final def adaptFromJournal(repr: PersistentRepr): PersistentRepr = {
+    val adapter = eventAdapters.get(repr.payload.getClass)
```

Comment:
Something is missing here. The stored format is typically the same for all events, e.g. String, Json, but there is no way to map that class to different adapters. I think something similar to the serializerId is needed, i.e. the id is stored and on the way back the adapter is looked up by id instead of by class.